### ISSUE TITLE
Optionally return attention weights in tf.keras.layers.BaseDenseAttention to allow us to visualize them

### ISSUE THREAD

**Issue by KieranLitschel** [tensorflow/tensorflow#44127]
**Describe the feature and the current behavior/state.**

Currently, when using attention in Keras, the weights are calculated in tf.keras.layers.BaseDenseAttention._apply_scores, and then discarded with only the score returned. It would be nice if the weights could be returned as they do in the "Neural machine translation with attention" TensorFlow Core tutorial (you can find the tutorial [here](https://www.tensorflow.org/tutorials/text/nmt_with_attention#write_the_encoder_and_decoder_model), see in the class BahdanauAttention in the tutorial how the attention_weights are calculated and then returned). This would be nice as it would allow us to visualize the scores like they do in the tutorial, which would help us interpret our models.

**Will this change the current api? How?**

Yes. As well as returning the regular output, it should also return the attention weights. 

**Who will benefit with this feature?**

Everyone who wants to better understand their model that uses attention. Being able to visualize the attention weights helps understand what the model is paying attention to in the input when predicting the output. 

**Any Other info.**

I'm surprised this is not already implemented in TensorFlow, so if there's any way I can achieve this without this change, please say, as I'm sure it will help others in the future too.

**Comment by KieranLitschel** [comment]
I just realized as self-attention is parameterless, another option is just to extract the inputs to attention from the model at test time and then calculate the weights using them. For example, if using Luong-style attention you'd extract the query and key and calculate it as:

`scores = tf.matmul(query, key, transpose_b=True)
weights = tf.nn.softmax(scores)`

Then the weights variable is the same that they calculate in the tutorial, so this can just be visualized to better understand what attention is actually doing. Maybe this is actually a better way to approach it instead of changing the API?

**Comment by fchollet** [comment]
This feature is available in TF Nightly and will be present in the next release: you can pass `return_attention_scores=True` in the `__call__` method of the dense attention layers. When True, the layer call will return a tuple `result, attention_scores`.
